lib/hanami/cli/commands/app/db/command.rb

```diff
@@ -30,32 +41,31 @@ def databases(app: false, slice: nil)
             end
 
             def database_for_app
-              Utils::Database[app]
+              build_database(app)
             end
 
             def database_for_slice(slice)
               slice = inflector.underscore(Shellwords.shellescape(slice)).to_sym
 
-              Utils::Database[app.slices[slice]]
+              build_database(app.slices[slice])
             end
 
             def all_databases
               slices = [app] + app.slices.with_nested
 
               slices_by_database_url = slices.each_with_object({}) { |slice, hsh|
-                next unless slice.container.providers.find_and_load_provider(:db)
-
-                slice.prepare :db
-                database_url = slice["db.gateway"].connection.uri
+                provider = slice.container.providers.find_and_load_provider(:db)
+                next unless provider
 
+                database_url = provider.source.send(:database_url)
```

Comment:
Avoid unnecessary connections to the database by reaching inside the provider source and calling its `#database_url` method (which we presume to be present). Right now `#database_url` is a private method, but we can make it public before this PR merges.

In addition, this method of accessing a provider object itself is not proper public API within dry-system. I think we need to streamline it:

- Have `Dry::System::ProviderRegistrar#[]` invoke `find_and_load_provider` so callers don't need to worry about that step (and to ensure consistent behaviour for provider registrars that aren't finalised)
- Classify `Dry::System::ProviderRegistrar#[]` as `@api public`
- Classify `Dry::System::Provider#source` as `@api public`
- Possibly even consider having `Dry::System::Provider` delegate missing methods to its source for nicer ergonomics, though I think we can skip this for now.

Why have I needed to do this?

Without this change, we had to establish a database connection just to get a slice's database_url, which we already know from config alone! This is wasteful and pointless. Worse still, we would actually connect multiple times to the same database for an app with multiple slices using the same database URL.

Why do we need to call this `provider.source.database_url`, though? This is to make sure we're using the "source of truth" for this database URL, which is the provider source itself, since it contains the logic to determine this URL when it eventually does connect.

It also makes testing this CLI command much harder (i.e. forcing us to have running database servers).

@solnic @flash-gordon @alassek — I'd be interested in your thoughts on this kind of provider usage. This feels reasonable to me, but it does make providers potentially "richer" objects in regular usage. I think this is safe though, since you couldn't confidently call methods against a provider's source unless you were already confident about what it should be.

An alternative would be to find some other home for this database_url logic, e.g. a static method somewhere in hanami-db itself. For this to work, though, we'd still need to send config from the provider to that method, since a user may choose to override the database_url directly in their own provider code. This is why I decided to stick with the provider as the source of truth and just use that directly.

Reply:
After a few days wait, I've gone ahead and merged this. I think it's a reasonable approach, and since this is all internal (i.e. no public API commitments), we can change it in the future if we need.

Reply:
@timriley I cannot see any problem with your approach. Is setting a URL is the only way to configure a DB connection?